Mathlib/LinearAlgebra/RootSystem/Finite/CanonicalBilinear.lean

```diff
@@ -132,6 +133,71 @@ lemma rootForm_root_self (j : ι) :
     P.RootForm (P.root j) (P.root j) = ∑ (i : ι), (P.pairing j i) * (P.pairing j i) := by
   simp [rootForm_apply_apply]
 
+theorem range_polarization_domRestrict_le_span_coroot :
+    LinearMap.range (P.Polarization.domRestrict (Submodule.span R (Set.range P.root))) ≤
+      (Submodule.span R (Set.range P.coroot)) := by
+  intro y hy
+  obtain ⟨x, hx⟩ := hy
+  rw [← hx, LinearMap.domRestrict_apply, Polarization_apply]
+  refine (mem_span_range_iff_exists_fun R).mpr ?_
+  use fun i => (P.toPerfectPairing x) (P.coroot i)
+  simp
+
 end CommRing
 
+theorem sum_mul_self_eq_zero_iff {ι : Type*} {R} [LinearOrderedCommRing R] (s : Finset ι)
+    (f : ι → R) : ∑ i ∈ s, f i * f i = 0 ↔ ∀ i ∈ s, f i = 0 := by
+  induction s using Finset.cons_induction with
+  | empty => simp
+  | cons i s his h =>
+    simp only [Finset.sum_cons, Finset.mem_cons, forall_eq_or_imp]
+    refine ⟨fun hc => ?_, fun hz => by simpa [hz.1] using h.mpr hz.2⟩
+    have hhi : f i * f i ≤ 0 := by
+      rw [← hc, le_add_iff_nonneg_right]
+      exact Finset.sum_nonneg fun i _ ↦ mul_self_nonneg (f i)
+    have hiz : f i * f i = 0 := (eq_of_le_of_le (mul_self_nonneg (f i)) hhi).symm
+    exact ⟨zero_eq_mul_self.mp hiz.symm, h.mp (by rw [← hc, hiz, zero_add])⟩
+
```

Comment:
This should be moved. I think possibly `Mathlib.Algebra.Order.BigOperators.Ring.Finset` might be a good home.